### PR TITLE
Lastopenedfile to socketserver

### PR DESCRIPTION
**Pull request by jeanawhou**
ticket: https://airtable.com/appfbr7vlSqIfZ9Sn/tblAmQkxs32m46J4N/viwURSVZRtP7uz0I8/recOFjSPHxDL5hm0Y?blocks=hide

### CODE REVIEW

File: main/listeners.js [hunk 1]
```diff
@@ -34,42 +34,28 @@ import {
...
-              'state-fetched',
-              fileURL,
-              flags,
-              numberOfWindows(),
-              win.fileURL,
-              processSwitches.serialise()
-            )
-          })
-        }
+  ipcMain.on('pls-fetch-state', function (event, id, lastFile, proMode) {
```
Comment (Quiescent) [inline main/listeners.js]: We should instead call out to the socket server from here.

File: main/modules/lastOpened.js [hunk 2]
```diff
@@ -1,48 +1,15 @@
...
-      return lastFileURL
-    })
-    .catch((error) => {
-      if (error.code === 'ENOENT') {
-        return null
-      }
-      log.error('Error finding last opened file.  Refusing to open it.', error)
-      return null
-    })
+const lastOpenedFile = () => {
```
Comment (Quiescent) [inline main/modules/lastOpened.js:3]: Use this method from the `pls-fetch-state` listener.

File: main/server/server.js [hunk 3]
```diff
@@ -844,6 +848,22 @@ const setupListeners = (port, userDataPath) => {
               () => 'Error while getting current backups'
             )
           }
+          case LAST_OPENED_FILE: {
+            return handlePromise(
+              () => 'Getting the last opened file',
+              () => lastOpenedFile,
```
Comment (Quiescent) [inline main/server/server.js]: You need to call `lastOpenedFile` or unwrap the lambda here.

File: src/app/containers/Main.js [hunk 4]
```diff
@@ -206,7 +206,7 @@ const Main = ({
       ipcRenderer.removeListener('state-fetched', stateFetchedListener)
     }
     ipcRenderer.on('state-fetched', stateFetchedListener)
-    ipcRenderer.send('pls-fetch-state', win.id, isInProMode)
+    ipcRenderer.send('pls-fetch-last-file', win.id, isInProMode)
```
Comment (Quiescent) [inline src/app/containers/Main.js]: This is what's preventing the file from opening.  We need the reply to `pls-fetch-state` to open a file :)

Comment (Quiescent) [reply src/app/containers/Main.js]: Oh wait, I see what you did later.  Reading further.

Comment (Quiescent) [reply src/app/containers/Main.js]: When you send on the renderer, it'll be heard on the main process.  It looks like you setup the listener for this ipc call in the renderer?

File: src/app/index.js [hunk 5]
```diff
@@ -307,6 +307,28 @@ ipcRenderer.on('redo', (event) => {
   store.dispatch(ActionCreators.redo())
 })
 
+ipcRenderer.on('pls-fetch-last-file', (event, id, isInProMode) => {
```
Comment (Quiescent) [inline src/app/index.js]: This listens for messages from the main process.

Comment (Quiescent) [reply src/app/index.js]: I think that you can avoid calling `pls-fetch-last-file` and then `pls-fetch-state` if you put the call to fetch the last opened file in the main process's listener.  It's fine to use the socket server client from the main process.

File: main/server/server.js [hunk 6]
```diff
@@ -844,6 +848,22 @@ const setupListeners = (port, userDataPath) => {
               () => 'Error while getting current backups'
             )
           }
+          case LAST_OPENED_FILE: {
+            return handlePromise(
+              () => 'Getting the last opened file',
+              () => lastOpenedFile(),
```
Comment (Quiescent) [inline main/server/server.js]: This task might take time, we should register it with the `statusManager`.
(Tasks registered with the `statusManager` will inform the renderer that the socket server is busy and also prevent Plottr from quitting.)

**Review by Quiescent — APPROVED** [review]
I'd like one small change before you merge (to use the `statusManager` in one more place).
But other than that, it's looking great.  Well done! 👍

**Comment by jeanawhou** [comment]
> I'd like one small change before you merge (to use the `statusManager` in one more place). But other than that, it's looking great. Well done! 👍

@Quiescent thank you so much for your teaching, guiding and advising me until i finish this task 🥲 🙇‍♀️

**Pull request merged**